File: 0282-Expression Add Operators/0282-Expression Add Operators.cpp
class Solution {
public:
    vector<string> addOperators(string num, int target) {
        vector<string> result;
        long currVal;
        for (int i = 1; i <= num.size(); ++i)
        {
            string path = num.substr(0, i);
            currVal = stol(path);
            if (path != to_string(currVal))
            {
                break;
            }
            dfs(num, i, currVal, currVal, ' ', path, target, result);
        }
        
        return result;
    }
    
private:
    void dfs(string num, int start, long preVal, long currVal, char preOp, string path, int target, vector<string>& result)
    {
        if (start == num.size())
        {
            if (currVal == target)
            {
                result.push_back(path);
            }
            return;
        }
        
        for (int i = start; i < num.size(); ++i)
        {
            string temp = num.substr(start, i - start + 1);
            long cv = stol(temp);
            if (temp != to_string(cv))
            {
                break;
            }
            dfs(num, i + 1, cv, currVal + cv, '+', path + '+' + temp, target, result);
            dfs(num, i + 1, cv, currVal - cv, '-', path + '-' + temp, target, result);
            long val;
            if (preOp == '+')
            {
                val = currVal - preVal + preVal * cv;
            }
            else if (preOp == '-')
            {
                val = currVal + preVal - preVal * cv;
            }
            else
            {
                val = currVal * cv;
            }
            dfs(num, i + 1, preVal * cv, val, preOp, path + '*' + temp, target, result);
        }
    }
};

File: 0201-0300/0282-Expression Add Operators/0282-Expression Add Operators.py
class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        if not num:
            return []
        result = []
        def dfs(start, path, curr, prev):
            if start == len(num):
                if curr == target:
                    result.append(path)
                return
            
            for i in range(start, len(num)):
                s = num[start: i + 1]
                n = int(s)
                if str(n) != s:
                    break
                    
                if start == 0:
                    dfs(i + 1, s, n, n)
                else:
                    dfs(i + 1, path + '+' + s, curr + n, n)
                    dfs(i + 1, path + '-' + s, curr - n, -n)
                    dfs(i + 1, path + '*' + s, curr - prev + prev*n, prev*n)
        
        dfs(0, '', 0, 0)
        return result
